### PR TITLE
Implement Cat.eager_subs for Slices

### PR DESCRIPTION
**Pull request by eb8680**
Blocking #299 

This PR adds the ability to substitute `Slice`s into the concatenation dimension of lazy `Cat` funsors.  by making `Cat.eager_subs` compute `Slice`s for each of its `parts`.

Tested:
- Added a test to `test_terms.py` comparing eager concatentation of `Tensor`s via `torch.cat` and lazy concatenations via `Cat.eager_subs`

### CODE REVIEW

**Comment by fritzo** [comment]
What is the motivation?

**Comment by eb8680** [comment]
> What is the motivating application?

Failing tests in #299

File: funsor/terms.py [hunk 1]
```diff
@@ -1371,6 +1371,28 @@ def eager_subs(self, subs):
                     return part(**{self.part_name: n})
                 n -= size
             assert False
+        elif isinstance(value, Slice):
+            start, stop, step = \
+                value.slice.start, value.slice.stop, value.slice.step
+            new_parts = []
+            pos = 0
+            for part in self.parts:
+                pdtype = part.inputs[self.part_name].dtype
```
Comment (fritzo) [inline funsor/terms.py]: nit: this would be a little clearer spelled
```py
psize = parts.inputs[self.part_name].size
```

File: funsor/torch.py [hunk 2]
```diff
@@ -607,12 +607,12 @@ def eager_cat_homogeneous(name, part_name, *parts):
...
         tensors.append(align_tensor(inputs, part).expand(shape))
-    if part_name != name:
-        del inputs[part_name]
+    del inputs[part_name]
 
     dim = 0
     tensor = torch.cat(tensors, dim=dim)
-    inputs[name] = bint(tensor.size(dim))
+    inputs = tuple(inputs.items())
+    inputs = OrderedDict(inputs[:dim] + ((name, bint(tensor.size(dim))),) + inputs[dim:])
```
Comment (fritzo) [inline funsor/torch.py]: This is overly complex since `dim` is always zero. How about simplifying to something like
```py
inputs = OrderedDict([(name, bint(tensor.size(dim))] + list(inputs.items()))
```
or
```py
inputs, trailing_inputs = OrderedDict([(name, bint(tensor.size(dim))]), inputs
inputs.update(trailing_inputs)
```

File: funsor/terms.py [hunk 3]
```diff
@@ -1371,6 +1371,28 @@ def eager_subs(self, subs):
                     return part(**{self.part_name: n})
                 n -= size
             assert False
+        elif isinstance(value, Slice):
+            start, stop, step = \
+                value.slice.start, value.slice.stop, value.slice.step
+            new_parts = []
+            pos = 0
+            for part in self.parts:
+                psize = part.inputs[self.part_name].dtype
```
Comment (fritzo) [inline funsor/terms.py]: You can even write `part.inputs[self.part_name].size`

**Comment by fritzo** [comment]
I have filed the torchvision bug upstream at https://github.com/pytorch/vision/issues/1726
